state.c

```diff
@@ -0,0 +1,337 @@
+    // If we get no user back, we have the new protocol
+    user_not_registered = signald_new_state(subscribed, "new protocol detected, get contacts", "user_not_registered", signald_new_protocol, signald_initialize_contacts);
+    // If we get an error, we have the old protocol
+    unexpected_error = signald_new_state(subscribed, "old protocol detected, get contacts", "unexpected_error", signald_old_protocol, signald_initialize_contacts);
+
+    /*
+     * Once the contact list comes back, we request the group list.
+     */
+    contacts = signald_new_state(user, "got contacts, get groups", "contact_list", signald_received_contact_list, signald_get_groups);
+    signald_add_transition(user_not_registered, "get_contacts", contacts, signald_received_contact_list, signald_get_groups);
+    signald_add_transition(unexpected_error, "get_contacts", contacts, signald_received_contact_list, signald_get_groups);
+
+    /*
+     * Group list has arrived, transition to running state.
+     */
+    running = signald_new_state(contacts, "running", "group_list", signald_received_group_list, NULL);
+
+    /*
+     * Normal even loop.  We transition from "running" back to "running", with
+     * various for the message types.
+     */
+    signald_add_transition(running, "message", running, signald_received_message, NULL);
+    signald_add_transition(running, "group_created", running, signald_get_groups, NULL);
+    signald_add_transition(running, "group_updated", running, signald_get_groups, NULL);
+    signald_add_transition(running, "group_list", running, signald_received_group_list, NULL);
+    signald_add_transition(running, "account_list", running, signald_received_account_list, NULL);
+
+    current = entrypoint;
+}
+
+gboolean
+signald_handle_message(SignaldAccount *sa, JsonObject *obj)
+{
+    const gchar *type = json_object_get_string_member(obj, "type");
+    SignaldStateTransition *transition = g_hash_table_lookup(current->transitions, type);
+
+    purple_debug_info(SIGNALD_PLUGIN_ID,
+                       "Initializing: In state '%s' received message type '%s'\n", 
+                       current->name,
+                       type);
+
+    if (transition == NULL) {
+        return FALSE;
+    }
+
+    if ((transition->handler != NULL) && ! transition->handler(sa, obj)) {
+        return FALSE;
+    }
+
+    purple_debug_info(SIGNALD_PLUGIN_ID,
+                      "Initializing: Transitioned from '%s' to '%s'\n", 
+                      current->name,
+                      transition->next->name);
+
+    if ((transition->next_message != NULL) && ! transition->next_message(sa, obj)) {
+        return FALSE;
+    }
+
+    current = transition->next;
+
+    if (g_hash_table_size(current->transitions) == 0) {
+        sa->initialized = TRUE;
+
+        purple_debug_info(SIGNALD_PLUGIN_ID, "Initializing: Initialization complete!\n");
```

Comment:
I should kill `SignaldAccount.initialized` entirely and probably emit an error, here, since there should be no terminal state anymore.